File: frontend/src/types/navigation.ts
/**
 * Represents a navigation route in the system with hierarchical support
 */
export interface NavigationRoute {
  /**
   * Unique identifier for the navigation route
   */
  id: string;
  
  /**
   * Display name shown in the navigation sidebar
   */
  name: string;
  
  /**
   * The route path used in URLs
   */
  route: string;
  
  /**
   * Icon identifier for the route (Material UI icon name)
   */
  icon?: string;
  
  /**
   * Description of the route
   */
  description?: string;
  
  /**
   * Order in which the route appears in navigation (lower numbers appear first)
   * @deprecated Use display_order instead for hierarchical navigation
   */
  order?: number;
  
  /**
   * Whether the route is visible in navigation
   */
  is_visible?: boolean;
  
  /**
   * ID of the user who created the route
   */
  creator_id: string;
  
  /**
   * Whether this is a system route that cannot be deleted
   */
  is_system_route?: boolean;
  
  /**
   * ID of the default component to display for this route
   */
  default_component_id?: string;
  
  /**
   * ID of the default page to display for this route
   */
  default_page_id?: string;
  
  /**
   * Whether the default component/page can be changed
   */
  can_change_default?: boolean;
  
  /**
   * Creation timestamp
   */
  created_at?: string;
  
  /**
   * Last update timestamp
   */
  updated_at?: string;
  
  // HIERARCHICAL FIELDS
  /**
   * ID of the parent navigation route (null for root level routes)
   */
  parent_id?: string;
  
  /**
   * Order within the parent group (lower numbers appear first)
   */
  display_order?: number;
  
  /**
   * Whether this section can be collapsed/expanded
   */
  is_collapsible?: boolean;
  
  /**
   * Whether this section is currently expanded
   */
  is_expanded?: boolean;
}

/**
 * Navigation route with children for tree structure
 */
export interface NavigationRouteTree extends NavigationRoute {
  /**
   * Child navigation routes
   */
  children?: NavigationRouteTree[];
  
  /**
   * Depth level in the tree (0 = root, 1 = child, etc.)
   */
  depth_level?: number;
}

/**
 * Data for moving a navigation route
 */
export interface NavigationRouteMove {
  /**
   * New parent ID (null to move to root level)
   */
  parent_id?: string;
  
  /**
   * New display order within the parent
   */
  display_order?: number;
}

/**
 * Data for batch updating navigation routes
 */
export interface NavigationRouteBatchUpdate {
  /**
   * ID of the route to update
   */
  id: string;
  
  /**
   * New parent ID
   */
  parent_id?: string;
  
  /**
   * New display order
   */
  display_order?: number;
  
  /**
   * New expanded state
   */
  is_expanded?: boolean;
}

/**
 * Navigation state for managing expanded/collapsed sections
 */
export interface NavigationState {
  /**
   * Set of route IDs that are currently expanded
   */
  expandedRoutes: Set<string>;
  
  /**
   * Set of route IDs that are currently collapsed
   */
  collapsedRoutes: Set<string>;
}

File: backend/app/schemas/navigation.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, UUID4, Field, validator
from uuid import UUID
from app.schemas.user import UserBase

class NavigationRouteBase(BaseModel):
    name: str
    route: str
    icon: Optional[str] = None
    description: Optional[str] = None
    order: Optional[int] = 0
    is_visible: Optional[bool] = True
    is_system_route: Optional[bool] = False
    default_component_id: Optional[str] = None
    default_page_id: Optional[UUID] = Field(default=None, description="Can be null to clear the default page")
    can_change_default: Optional[bool] = False
    
    # Hierarchical fields
    parent_id: Optional[str] = None
    display_order: Optional[int] = 0
    is_collapsible: Optional[bool] = True
    is_expanded: Optional[bool] = True
    
    @validator('default_page_id', pre=True)
    def validate_default_page_id(cls, v):
        if v is None or v == "":
            return None
        return v

class NavigationRouteCreate(NavigationRouteBase):
    @validator('default_page_id', pre=True)
    def validate_default_page_id(cls, v):
        if v is None or v == "":
            return None
        return v
from pydantic import Field, validator

class NavigationRouteUpdate(BaseModel):
    name: Optional[str] = None
    route: Optional[str] = None
    icon: Optional[str] = None
    description: Optional[str] = None
    order: Optional[int] = None
    is_visible: Optional[bool] = None
    default_component_id: Optional[str] = None
    default_page_id: Optional[UUID] = Field(default=None, description="Can be null to clear the default page")
    can_change_default: Optional[bool] = None
    parent_id: Optional[str] = None
    display_order: Optional[int] = None
    is_collapsible: Optional[bool] = None
    is_expanded: Optional[bool] = None
    # Note: is_system_route should not be updatable
    
    @validator('default_page_id', pre=True)
    def validate_default_page_id(cls, v):
        if v is None or v == "":
            return None
        return v
    
    class Config:
        # Allow conversion from string to UUID for default_page_id
        json_encoders = {
            UUID: lambda v: str(v)
        }
        # Allow null values for default_page_id
        extra = "allow"
        validate_assignment = True

class NavigationRouteResponse(NavigationRouteBase):
    id: str
    creator_id: str
    created_at: Optional[str] = None
    updated_at: Optional[str] = None

    class Config:
        from_attributes = True

class NavigationRouteDetailResponse(NavigationRouteResponse):
    creator: Optional[dict] = None
    pages: Optional[List[str]] = None

class NavigationRouteTree(NavigationRouteResponse):
    children: Optional[List['NavigationRouteTree']] = []
    depth_level: Optional[int] = 0

class NavigationRouteMove(BaseModel):
    parent_id: Optional[str] = None
    display_order: Optional[int] = None

class NavigationRouteBatchUpdate(BaseModel):
    id: str
    parent_id: Optional[str] = None
    display_order: Optional[int] = None
    is_expanded: Optional[bool] = None

class NavigationRouteListResponse(BaseModel):
    routes: List[NavigationRouteResponse]
    total: int

# Enable forward references for NavigationRouteTree
NavigationRouteTree.model_rebuild()
